hbase-examples/src/main/java/org/apache/hadoop/hbase/coprocessor/example/row/stats/RowStatisticsCompactionObserver.java

```diff
@@ -0,0 +1,321 @@
+        if (shouldCreateNamespace) {
+          NamespaceDescriptor nd = NamespaceDescriptor.create(NAMESPACE).build();
+          try {
+            admin.createNamespace(nd);
+          } catch (IOException e) {
+            LOG.error("Failed to create namespace {}", NAMESPACE, e);
+          }
+        }
+        ColumnFamilyDescriptor cfd = ColumnFamilyDescriptorBuilder.newBuilder(CF).setMaxVersions(25)
+          .setTimeToLive((int) Duration.ofDays(7).toSeconds()).build();
+        TableDescriptor td =
+          TableDescriptorBuilder.newBuilder(NAMESPACED_TABLE_NAME).setColumnFamily(cfd).build();
+        LOG.info("Creating table {}", NAMESPACED_TABLE_NAME);
+        try {
+          admin.createTable(td);
+        } catch (IOException e) {
+          LOG.error("Failed to create table {}", NAMESPACED_TABLE_NAME, e);
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get Connection or Admin. Cannot determine if table {} exists.",
+        NAMESPACED_TABLE_NAME, e);
+    }
+  }
+
+  @Override
+  public InternalScanner preCompact(ObserverContext<RegionCoprocessorEnvironment> context,
+    Store store, InternalScanner scanner, ScanType scanType, CompactionLifeCycleTracker tracker,
+    CompactionRequest request) {
+    if (RowStatisticsUtil.isInternalTable(store.getTableName())) {
+      LOG.debug("Region {} belongs to an internal table {}, so no row statistics will be recorded",
+        store.getRegionInfo().getRegionNameAsString(), store.getTableName().getNameAsString());
+      return scanner;
+    }
+    int blocksize = store.getColumnFamilyDescriptor().getBlocksize();
+    boolean isMajor = request.isMajor();
+    RowStatisticsImpl stats;
+    if (isMajor) {
+      stats = new RowStatisticsImpl(store.getTableName().getNameAsString(),
+        store.getRegionInfo().getEncodedName(), store.getColumnFamilyName(), blocksize,
+        maxCacheSize, true);
+    } else {
+      stats = new RowStatisticsImpl(store.getTableName().getNameAsString(),
+        store.getRegionInfo().getEncodedName(), store.getColumnFamilyName(), blocksize,
+        maxCacheSize, false);
+    }
+    return new RowStatisticsScanner(scanner, isMajor, stats, context.getEnvironment());
+  }
+
+  private class RowStatisticsScanner implements InternalScanner, Shipper {
+
+    private final InternalScanner scanner;
+    private final Shipper shipper;
+    private final boolean isMajor;
+    private final RowStatisticsImpl rowStatistics;
+    private final RegionCoprocessorEnvironment regionEnv;
+    private RawCellBuilder cellBuilder;
+    private Cell lastCell;
+
+    public RowStatisticsScanner(InternalScanner scanner, boolean isMajor,
+      RowStatisticsImpl rowStatistics, RegionCoprocessorEnvironment regionEnv) {
+      this.scanner = scanner;
+      if (scanner instanceof Shipper) {
+        this.shipper = (Shipper) scanner;
+      } else {
+        this.shipper = null;
+      }
+      this.isMajor = isMajor;
```

Comment:
nit: trim off one constructor argument and read `isMajor` from the `rowStatistics` value.